File: src/Python/src/vladimir_reshetnikov/data_structures/hamt/persistent_hash_multimap.py
"""Set-valued persistent hash multimap over the public CHAMP map and set."""

from __future__ import annotations

from collections.abc import Iterable, Iterator
from dataclasses import dataclass
from typing import Generic, TypeVar

from .hash_policy import HashPolicy, default_hash_policy
from .persistent_hamt import PersistentHashMap, PersistentHashSet

K = TypeVar("K")
V = TypeVar("V")

_INT64_MAX = (1 << 63) - 1


@dataclass(frozen=True, slots=True)
class HashMultimapEntry(Generic[K, V]):
    """One stored representative pair."""

    key: K
    value: V


@dataclass(frozen=True, slots=True)
class HashMultimapKeyResult(Generic[K]):
    """Presence-safe stored key lookup; ``key`` may itself be ``None``."""

    found: bool
    key: K | None = None


@dataclass(frozen=True, slots=True)
class HashMultimapValuesResult(Generic[V]):
    """Presence-safe lookup of a non-empty value group."""

    found: bool
    values: PersistentHashSet[V]


class PersistentHashMultimap(Generic[K, V]):
    """Immutable set-valued hash multimap with independent domain policies.

    The outer CHAMP map stores no empty value groups. Both domains retain their first
    representatives, and duplicate addition is an exact object-identity no-op.
    """

    __slots__ = ("_groups", "_pair_count", "_value_policy")

    def __init__(
        self,
        groups: PersistentHashMap[K, PersistentHashSet[V]],
        value_policy: HashPolicy[V],
        pair_count: int,
    ) -> None:
        self._groups = groups
        self._value_policy = value_policy
        self._pair_count = pair_count

    @classmethod
    def empty(
        cls,
        key_policy: HashPolicy[K] | None = None,
        value_policy: HashPolicy[V] | None = None,
    ) -> PersistentHashMultimap[K, V]:
        """Return an empty multimap retaining both exact policy objects."""

        effective_keys = default_hash_policy() if key_policy is None else key_policy
        effective_values = default_hash_policy() if value_policy is None else value_policy
        groups: PersistentHashMap[K, PersistentHashSet[V]] = PersistentHashMap.empty(effective_keys)
        return cls(groups, effective_values, 0)

    @classmethod
    def from_items(
        cls,
        items: Iterable[tuple[K, V]],
        key_policy: HashPolicy[K] | None = None,
        value_policy: HashPolicy[V] | None = None,
    ) -> PersistentHashMultimap[K, V]:
        """Build from pairs, ignoring duplicates under the retained policies."""

        if items is None:
            raise TypeError("items must be iterable.")
        result = cls.empty(key_policy, value_policy)
        for key, value in items:
            result = result.add(key, value)
        return result

    @property
    def key_count(self) -> int:
        return self._groups.size

    @property
    def pair_count(self) -> int:
        return self._pair_count

    @property
    def is_empty(self) -> bool:
        return self._pair_count == 0

    @property
    def key_policy(self) -> HashPolicy[K]:
        return self._groups.policy

    @property
    def value_policy(self) -> HashPolicy[V]:
        return self._value_policy

    def __len__(self) -> int:
        return self._pair_count

    def __bool__(self) -> bool:
        return not self.is_empty

    def contains_key(self, key: K) -> bool:
        return self._groups.contains_key(key)

    def contains(self, key: K, value: V) -> bool:
        group = self._groups.get_entry(key)
        return group is not None and group.value.contains(value)

    def try_get_key(self, key: K) -> HashMultimapKeyResult[K]:
        """Return the first stored key representative without conflating ``None`` and absence."""

        group = self._groups.get_entry(key)
        return (
            HashMultimapKeyResult(False)
            if group is None
            else HashMultimapKeyResult(True, group.key)
        )

    def get_values(self, key: K) -> PersistentHashSet[V]:
        """Return the represented group or a value-policy-preserving empty set."""

        group = self._groups.get_entry(key)
        return PersistentHashSet.empty(self._value_policy) if group is None else group.value

    def try_get_values(self, key: K) -> HashMultimapValuesResult[V]:
        group = self._groups.get_entry(key)
        return (
            HashMultimapValuesResult(False, PersistentHashSet.empty(self._value_policy))
            if group is None
            else HashMultimapValuesResult(True, group.value)
        )

    def add(self, key: K, value: V) -> PersistentHashMultimap[K, V]:
        """Add one pair or return this object when it already exists."""

        group = self._groups.get_entry(key)
        if group is None:
            if self._pair_count == _INT64_MAX:
                raise OverflowError("The operation would exceed the signed 64-bit pair count.")
            values = PersistentHashSet.empty(self._value_policy).add(value)
            return PersistentHashMultimap(
                self._groups.add(key, values), self._value_policy, self._pair_count + 1
            )

        added = group.value.try_add(value)
        if not added.added:
            return self
        if self._pair_count == _INT64_MAX:
            raise OverflowError("The operation would exceed the signed 64-bit pair count.")
        return PersistentHashMultimap(
            self._groups.put(group.key, added.value),
            self._value_policy,
            self._pair_count + 1,
        )

    def remove(self, key: K, value: V) -> PersistentHashMultimap[K, V]:
        """Remove one pair and contract a group when its final value disappears."""

        group = self._groups.get_entry(key)
        if group is None:
            return self
        removed = group.value.try_remove(value)
        if removed is None:
            return self
        groups = (
            self._groups.remove(group.key)
            if removed.set.is_empty
            else self._groups.put(group.key, removed.set)
        )
        return PersistentHashMultimap(groups, self._value_policy, self._pair_count - 1)

    def remove_key(self, key: K) -> PersistentHashMultimap[K, V]:
        """Remove one complete key class and all of its pairs."""

        removed = self._groups.try_remove_entry(key)
        if removed is None:
            return self
        return PersistentHashMultimap(
            removed.map,
            self._value_policy,
            self._pair_count - removed.entry.value.size,
        )

    def clear(self) -> PersistentHashMultimap[K, V]:
        return self if self.is_empty else self.empty(self.key_policy, self._value_policy)

    def keys(self) -> Iterator[K]:
        return self._groups.keys()

    def __iter__(self) -> Iterator[HashMultimapEntry[K, V]]:
        for group in self._groups:
            for value in group.value:
                yield HashMultimapEntry(group.key, value)

    def shares_root_with(self, other: PersistentHashMultimap[K, V]) -> bool:
        """Return whether the two facades reference the same outer CHAMP root."""

        return self._groups.shares_root_with(other._groups)

    def validate_structure(self) -> bool:
        """Recompute counts, policy propagation, and the no-empty-group invariant."""

        pairs = 0
        for group in self._groups:
            if group.value.is_empty or group.value.policy is not self._value_policy:
                return False
            pairs += group.value.size
            if pairs > _INT64_MAX:
                return False
        return pairs == self._pair_count and (pairs == 0) == self._groups.is_empty


__all__ = [
    "HashMultimapEntry",
    "HashMultimapKeyResult",
    "HashMultimapValuesResult",
    "PersistentHashMultimap",
]

File: src/Kotlin/Hamt/src/durable7/hamt/PersistentHashMultimap.kt
/*
 * Persistent set-valued hash multimap.
 *
 * Maps each key to a nonempty value set, so adding a present pair is a no-op and removing a key's
 * last value removes the key itself. Distinct-key and pair cardinalities are tracked separately,
 * because neither can be derived from the other in constant time.
 */
package durable7.hamt

/** One key-value pair as the multimap presents it. */
public data class MultimapEntry<K, V>(public val key: K, public val value: V)

/** Key and pair counts returned by a successful structural audit. */
public data class PersistentHashMultimapStatistics(
    public val keyCount: Int,
    public val pairCount: Long,
)

/** Immutable set-valued multimap retaining independent key and value policies. */
public class PersistentHashMultimap<K, V> private constructor(
    private val groups: PersistentHashMap<K, PersistentHashSet<V>>,
    public val valuePolicy: HashPolicy<V>,
    public val pairCount: Long,
) : Iterable<MultimapEntry<K, V>> {
    public companion object {
        /** An empty multimap. The key and value policies are independent and both are retained by identity. */
        public fun <K, V> empty(
            keyPolicy: HashPolicy<K> = defaultHashPolicy(),
            valuePolicy: HashPolicy<V> = defaultHashPolicy(),
        ): PersistentHashMultimap<K, V> =
            PersistentHashMultimap(PersistentHashMap.empty(keyPolicy), valuePolicy, 0L)

        /** A multimap holding every pair. Duplicate pairs collapse, since the values for a key form a set. */
        public fun <K, V> from(
            entries: Iterable<Pair<K, V>>,
            keyPolicy: HashPolicy<K> = defaultHashPolicy(),
            valuePolicy: HashPolicy<V> = defaultHashPolicy(),
        ): PersistentHashMultimap<K, V> {
            var result = empty<K, V>(keyPolicy, valuePolicy)
            for ((key, value) in entries) result = result.add(key, value)
            return result
        }
    }

    /** The hash policy used for keys. */
    public val keyPolicy: HashPolicy<K> get() = groups.policy
    /** Number of distinct keys, which is fewer than [pairCount] whenever any key has several values. */
    public val keyCount: Int get() = groups.size
    /** Whether the multimap holds no pairs. */
    public val isEmpty: Boolean get() = pairCount == 0L

    /**
     * Whether [key] has at least one value. A key never remains with an empty group: removing its last value
     * removes the key.
     */
    public fun containsKey(key: K): Boolean = groups.containsKey(key)
    /** Whether the pair is present. */
    public fun contains(key: K, value: V): Boolean = groups.getEntry(key)?.value?.contains(value) == true
    /**
     * The *stored* key representative equal to [key], or `null` when absent. The first representative of each
     * equivalence class is the one kept, so this need not be the instance passed in.
     */
    public fun actualKey(key: K): K? = groups.getEntry(key)?.key
    /**
     * The values for [key] as a set, or `null` when the key is absent. Never an empty set: absence and emptiness
     * are the same state here.
     */
    public fun valuesFor(key: K): PersistentHashSet<V>? = groups.getEntry(key)?.value

    /**
     * A multimap with the pair added. Returns the receiver unchanged when the pair is already present, since the
     * values for a key form a set.
     */
    public fun add(key: K, value: V): PersistentHashMultimap<K, V> {
        val indexed = groups.getEntry(key)
        if (indexed == null) {
            val group = PersistentHashSet.empty<V>(valuePolicy).add(value)
            return PersistentHashMultimap(groups.add(key, group), valuePolicy, Math.addExact(pairCount, 1L))
        }
        val added = indexed.value.tryAdd(value)
        if (!added.added) return this
        return PersistentHashMultimap(
            groups.put(indexed.key, added.value),
            valuePolicy,
            Math.addExact(pairCount, 1L),
        )
    }

    /**
     * A multimap without the pair. Removing a key's last value removes the key itself, so no empty group is left
     * behind. Returns the receiver unchanged when the pair is absent.
     */
    public fun remove(key: K, value: V): PersistentHashMultimap<K, V> {
        val indexed = groups.getEntry(key) ?: return this
        val removed = indexed.value.tryRemove(value) ?: return this
        val nextGroups = if (removed.set.isEmpty) groups.remove(indexed.key) else groups.put(indexed.key, removed.set)
        return PersistentHashMultimap(nextGroups, valuePolicy, pairCount - 1L)
    }

    /** A multimap without [key] and all of its values. Returns the receiver unchanged when the key is absent. */
    public fun removeKey(key: K): PersistentHashMultimap<K, V> {
        val indexed = groups.getEntry(key) ?: return this
        return PersistentHashMultimap(groups.remove(indexed.key), valuePolicy, pairCount - indexed.value.size.toLong())
    }

    /** An empty multimap retaining both policies. */
    public fun clear(): PersistentHashMultimap<K, V> =
        if (isEmpty) this else empty(keyPolicy, valuePolicy)

    /** Receiver-policy union; the complete argument is normalized while enumerating it. */
    public fun union(other: PersistentHashMultimap<K, V>): PersistentHashMultimap<K, V> {
        var result = this
        for (entry in other) result = result.add(entry.key, entry.value)
        return result
    }

    /** The pairs present in both multimaps, compared under the receiver's policies. */
    public fun intersect(other: PersistentHashMultimap<K, V>): PersistentHashMultimap<K, V> {
        val normalized = clear().union(other)
        var result = clear()
        for (entry in this) if (normalized.contains(entry.key, entry.value)) result = result.add(entry.key, entry.value)
        return result
    }

    /** The pairs of this multimap absent from [other], compared under the receiver's policies. */
    public fun except(other: PersistentHashMultimap<K, V>): PersistentHashMultimap<K, V> {
        val normalized = clear().union(other)
        var result = this
        for (entry in normalized) result = result.remove(entry.key, entry.value)
        return result
    }

    /** Every pair, as a flat list. */
    public fun toList(): List<MultimapEntry<K, V>> = iterator().asSequence().toList()

    /**
     * Walk the whole multimap and check its invariants - the cached pair count, the key count, and that no group is
     * empty - returning its shape measurements. A defensive audit for tests, not a routine operation.
     *
     * @throws IllegalStateException on the first violation found.
     */
    public fun validateStructure(): PersistentHashMultimapStatistics {
        var counted = 0L
        var countedKeys = 0
        for (group in groups) {
            check(!group.value.isEmpty) { "PersistentHashMultimap stores an empty group." }
            counted = Math.addExact(counted, group.value.size.toLong())
            countedKeys += 1
        }
        check(countedKeys == keyCount && counted == pairCount) { "PersistentHashMultimap counts disagree." }
        return PersistentHashMultimapStatistics(keyCount, pairCount)
    }

    override fun iterator(): Iterator<MultimapEntry<K, V>> = sequence {
        for (group in groups) for (value in group.value) yield(MultimapEntry(group.key, value))
    }.iterator()
}

/** Left, right, and pair counts returned by a successful structural audit. */
public data class PersistentRelationStatistics(
    public val leftCount: Int,
    public val rightCount: Int,
    public val pairCount: Long,
)

/** Bidirectional persistent many-to-many relation with exact inverse indexes. */
public class PersistentRelation<L, R> private constructor(
    private val forward: PersistentHashMultimap<L, R>,
    private val reverse: PersistentHashMultimap<R, L>,
) : Iterable<Pair<L, R>> {
    public companion object {
        /** An empty relation. The left and right policies are independent and both are retained by identity. */
        public fun <L, R> empty(
            leftPolicy: HashPolicy<L> = defaultHashPolicy(),
            rightPolicy: HashPolicy<R> = defaultHashPolicy(),
        ): PersistentRelation<L, R> = PersistentRelation(
            PersistentHashMultimap.empty(leftPolicy, rightPolicy),
            PersistentHashMultimap.empty(rightPolicy, leftPolicy),
        )
    }

    /** Number of related pairs. */
    public val pairCount: Long get() = forward.pairCount
    /** Number of distinct left values that relate to something. */
    public val leftCount: Int get() = forward.keyCount
    /** Number of distinct right values that relate to something. */
    public val rightCount: Int get() = reverse.keyCount
    /** Whether the relation holds no pairs. */
    public val isEmpty: Boolean get() = forward.isEmpty

    /** Whether [left] and [right] are related. */
    public fun contains(left: L, right: R): Boolean = forward.contains(left, right)
    /** The right values related to [left], or `null` when it relates to nothing. Never an empty set. */
    public fun rightsFor(left: L): PersistentHashSet<R>? = forward.valuesFor(left)
    /**
     * The left values related to [right], or `null` when it relates to nothing. Costs the same as [rightsFor],
     * since both directions are indexed.
     */
    public fun leftsFor(right: R): PersistentHashSet<L>? = reverse.valuesFor(right)

    /**
     * A relation with the pair added to both directions. Returns the receiver unchanged when the pair is already
     * present.
     */
    public fun add(left: L, right: R): PersistentRelation<L, R> =
        if (contains(left, right)) this
        else PersistentRelation(forward.add(left, right), reverse.add(right, left))

    /** A relation without the pair, removed from both directions. Returns the receiver unchanged when absent. */
    public fun remove(left: L, right: R): PersistentRelation<L, R> =
        if (!contains(left, right)) this
        else PersistentRelation(forward.remove(left, right), reverse.remove(right, left))

    /** A relation without [left] and every pair it participates in. */
    public fun removeLeft(left: L): PersistentRelation<L, R> {
        val rights = rightsFor(left)?.toList() ?: return this
        var result = this
        for (right in rights) result = result.remove(left, right)
        return result
    }

    /** A relation without [right] and every pair it participates in. */
    public fun removeRight(right: R): PersistentRelation<L, R> {
        val lefts = leftsFor(right)?.toList() ?: return this
        var result = this
        for (left in lefts) result = result.remove(left, right)
        return result
    }

    /** An empty relation retaining both policies. */
    public fun clear(): PersistentRelation<L, R> =
        if (isEmpty) this else PersistentRelation(forward.clear(), reverse.clear())

    /**
     * The relation with its two sides exchanged. O(1): both directions are already indexed, so this swaps them
     * rather than rebuilding.
     */
    public fun inverse(): PersistentRelation<R, L> = PersistentRelation(reverse, forward)

    /**
     * Walk both directions and check that they agree - matching pair counts, and every forward pair present in
     * reverse - returning shape measurements. A defensive audit for tests, not a routine operation.
     *
     * @throws IllegalStateException on the first violation found.
     */
    public fun validateStructure(): PersistentRelationStatistics {
        forward.validateStructure()
        reverse.validateStructure()
        check(forward.pairCount == reverse.pairCount) { "PersistentRelation pair counts disagree." }
        for (entry in forward) check(reverse.contains(entry.value, entry.key)) { "PersistentRelation inverse is missing." }
        for (entry in reverse) check(forward.contains(entry.value, entry.key)) { "PersistentRelation forward index is missing." }
        return PersistentRelationStatistics(leftCount, rightCount, pairCount)
    }

    override fun iterator(): Iterator<Pair<L, R>> = forward.asSequence().map { it.key to it.value }.iterator()
}
